pkg/main.go

```diff
@@ -0,0 +1,94 @@
+package pkg
+
+import (
+	"time"
+
+	"github.com/v3io/registry-creds-handler/pkg/registry"
+	"github.com/v3io/registry-creds-handler/pkg/util"
+
+	"github.com/nuclio/errors"
+	"github.com/nuclio/logger"
+	"k8s.io/client-go/kubernetes"
+)
+
+type Handler struct {
+	logger        logger.Logger
+	kubeClientSet *kubernetes.Clientset
+	ecr           *registry.ECR
+	refreshRate   time.Duration
+	registryKind  string
+}
+
+func NewHandler(logger logger.Logger,
+	kubeClientSet *kubernetes.Clientset,
+	ecr *registry.ECR,
+	refreshRate int64,
+	registryKind string) (*Handler, error) {
+
+	return &Handler{
+		logger:        logger.GetChild("handler"),
+		kubeClientSet: kubeClientSet,
+		ecr:           ecr,
+		refreshRate:   time.Duration(refreshRate) * time.Minute,
+		registryKind:  registryKind,
+	}, nil
+}
+
+func (h *Handler) Run() error {
+	h.logger.Info("Handler starting...")
+	switch h.registryKind {
+	case registry.ECRRegistryKind:
+		err := h.createOrUpdateECRSecret()
+		if err != nil {
+			return errors.Wrap(err, "Failed to create or update ECR token secret")
+		}
+
+		// should never return
+		err = util.SyncSecret(h.kubeClientSet, h.refreshRate, h.ecr.Namespace, h.ecr.SecretName, h.createOrUpdateECRSecret)
+		if err != nil {
+			return errors.Wrap(err, "Failed to sync secret")
+		}
+	default:
+		return errors.New("Received unsupported registry kind")
+	}
+	return errors.New("Handler exited unexpectedly")
+}
+
+func (h *Handler) createOrUpdateECRSecret() error {
```

Comment:
no need to be specific about the kind. you get an interface to use which exposes a functionallity. you dont care here about the concrete